python/examples/core/nevergrad_parallel_utils.py

```diff
@@ -17,166 +17,52 @@
-      # Join at least one process if nothing is available.
-      processes_joined = join_at_least_one_process(self.tasks)
-      assert len(processes_joined) > 0, "no processes were joined"
-      for process_idx in processes_joined:
-        throughput = tell_joined_process(self.tasks, process_idx, scheduler,
-                                         optimizer, throughputs, parsed_args)
-        throughput = int(throughput)
-        best = throughput if throughput > best else best
-        self.tasks[process_idx] = None
-
-    # We are sure there is at least one empty slot.
-    return self.tasks.index(None), best
+    self.throughputs = throughputs
 
+def cpu_count():
+  """Return the number of CPUs on which processes can be scheduled."""
+  return len(os.sched_getaffinity(0))
 
 def compile_and_run_checked_mp(problem: ProblemInstance, \
-                               lock: RWLock,
-                               compilation_cpu: int,
-                               benchmark_cpu_min: int,
-                               benchmark_cpu_max: int,
-                               scheduler: NGSchedulerInterface,
                                proposal,
-                               n_iters: int,
-                               timeout_compile: float,
-                               timeout_benchmark: float,
-                               ipc_dict: dict):
+                               scheduler: NGSchedulerInterface):
   """Entry point to compile and run while catching and reporting exceptions.
 
-  This is run in interruptible multiprocess mode.
+  This is run in interruptible multiprocess mode. Result are returned from this
+  process via `result_queue`.
 
-  ipc_dict must be provided, and it is used to return information across the
-  root / children process boundary:
-    - 'throughputs': the measured throughputs.
-    - 'success': the return status.
+  Before starting compilation, acquire a lock in read mode. Wait for a certain
+  amount of time and kill the process if compilation is still running. The
+  compilation is always run on a single CPU. (The process is alrady pinned to a
+  single CPU when entering this function.)
 
-  Acquire lock in read mode and spawn a compile thread, joined with a 
-  `timeout_compile` timeout to perform the compile part of the job. 
-  The compilation is always run on a single CPU, set by a taskset command.
-
-  Acquire lock in write mode and spawn a benchmark thread, joined with a 
-  `timeout_benchmark` timeout to perform the benchmark part of the job. 
-  The benchmark may be run on multiple CPUs, set by a taskset command.
+  Before starting a benchmark, acquire a lock in write mode. Wait for a certain
+  amount of time and kill the process if the benchmark is still running. The
+  benchmark may be run on multiple CPUs.
 
   Upon SIGTERM triggered by the parent, locks are simply dropped on the floor
   and everything stops.
   """
+
+  global results_queue, parsed_args, rw_locks
```

Comment:
do we need these to be globals?

Reply:
`results_queue` and `parsed_args` could be passed each time we enqueue a search job. Setting it up during process initialization and storing them in global variables is a bit more efficient.

The `rw_locks`, we actually need a global variable. I ran into a weird bug where pool processes kept crashing. `RWLock` should pickle according to the documentation, but there must be a bug in their implementation somewhere. When setting up stuff during process initialization, we can pass stuff that does not pickle.

So since we need global variables anyway, I'd keep the others as well for efficiency reasons.